compiler/rustc_const_eval/src/const_eval/mod.rs

```diff
@@ -39,6 +39,87 @@ pub(crate) fn const_caller_location(
     ConstValue::Scalar(Scalar::from_pointer(loc_place.ptr.into_pointer_or_addr().unwrap(), &tcx))
 }
 
+pub(crate) fn const_type_id<'tcx>(
```

Comment:
I think I'd prefer this helper to be in the `interpret` module, not `const_eval` -- it is needed for all interpreter instances after all, `const_eval` is meant for specifically the CTFE instance of the general interpreter infrastruture.

Reply:
You're also creating your own `ecx` here, and returning this as a `ConstValue` -- why all that?

I guess an alternative might be to treat this like `vtable_allocation`?

Reply:
`type_name::alloc_type_name` also seems similar in spirit?

Reply:
Oooh I see, I was modelling it after `const_caller_location` but that's only here because it's a query instead of an intrinsic. Should it have its own module under `interpret::intrinsics` like `caller_location` does?

Reply:
Yeah, note in particular that the `caller_location` implementation in `interpret` does *not* call the `const_eval` code. It is not a nullary intrinsic so it takes slightly different paths. Though probably the `const_caller_location` query should be made more like the vtable query, which is not inside `const_eval` either.

> Should it have its own module under interpret::intrinsics like caller_location does?

Yeah I guess that would make sense.
It would almost make sense to move these queries for intrinsics that are implemented via interpreter stuff but also used for codegen in a separate module `rustc_const_eval::intrinsics`, i.e., outside `interpret`. @oli-obk any thoughts?

Reply:
Looking at it, `caller_location` actually has a `impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M>` block so having that inside `interpret` makes sense. `type_name`, OTOH, doesn't really do anything with the interpreter, it is very much like the `vtable` code (but the two are in totally different locations, which we should probably fix). "give me the vtable for type T" is not technically an intrinsic but not very different, either, is it?

Reply:
I think cleaning this up is orthogonal.

Except for `mplace_field` we don't even need an `InterpCx` and can just create and intern the `Allocation`s directly. If we added more convenience operations for writing data of a specific struct to an `Allocation` we can probably stop using `InterpCx` for `caller_location`, too.

Imo we should just merge this PR and then work on cleaning these up now that we have multiple intrinsics to clean up in a similar way.

Reply:
I agree cleaning up the existing stuff is a separate task.

But we shouldn't have anything load-bearing inside the `const_eval` module, so I do think that this code should be moved somewhere else before landing. I don't care strongly which of the existing precedent it follows, but *none* of it has load-bearing code in const_eval, so let's not add yet another different style of implementing such an intrinsic. :)